weather_test.go

```diff
@@ -1,26 +1,57 @@
 package weather_test
 
 import (
-	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
 	"os"
 	"testing"
 	"weather"
 )
 
-func TestGet(t *testing.T) {
-	var conditions []byte
+func TestConstructUrl(t *testing.T) {
 
-	token := os.Getenv("WEATHERAPP_TOKEN")
-	if len(token) == 0 {
-		fmt.Printf("please set a weatherapp token\n")
-		os.Exit(2)
+	token := "foo"
+	location := weather.LocationFromArgs([]string{"PATH", "rio", "de", "janeiro"})
+	want := "https://api.openweathermap.org/data/2.5/weather?q=rio%20de%20janeiro&appid=foo"
+
+	got := weather.BuildURL(token, location)
+
+	if want != got {
+		t.Fatalf("want: %q got: %q", want, got)
 	}
+}
+
+func TestGetWeather(t *testing.T) {
 
-	conditions, err := weather.Get("london", token)
+	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		file, err := os.Open("testdata/weather.json")
+		if err != nil {
+			t.Fatal(err)
+		}
+		io.Copy(w, file)
+	}))
+	defer ts.Close()
+
+	client := ts.Client()
+	res, err := client.Get(ts.URL)
```

Comment:
This is all right, but the problem is we're not really testing any HTTP-handling code in our package. We're doing all the HTTP in the test, and the result is just some `io.Reader` containing the JSON. Well, if that's all we're doing, there's a much easier way to get that reader: just pass the result of `os.Open` to `weather.Get` directly!

It's always good to test as much of your real code path as you can. Here, that would mean making real HTTP calls to a real server: but to the `ts` server rather than the real API. You can use the 'client object' pattern to do this.